### PR TITLE
DDCE-5816: Scala 3 Upgrade

### PR DESCRIPTION
**Pull request by SachaMcCormack**
1. Upgraded to Scala 3
2. Reduced coverage to 95% after adding tests

### CODE REVIEW

**Comment by lmugnai** [comment]
Ran locally - PASS
Screenshot 2024-11-06 at 11 57 29

**Comment by lmugnai** [comment]
Ran ATs locally - PASS
Screenshot 2024-11-06 at 13 52 33

File: test/controllers/AccountHomeControllerSpec.scala [hunk 1]
```diff
@@ -102,7 +102,7 @@ class AccountHomeControllerSpec extends SpecBase {
 
       when(
         mockBackEndConnector.applicationSummaries(any())
-      ) thenReturn Future
+      ) `thenReturn` Future
```
Comment (lmugnai) [inline test/controllers/AccountHomeControllerSpec.scala:105]: I prefer using .thenReturn() over backticks. I've posted a comment in our slack dev channel to see what people say.

File: app/forms/AgentCompanyDetailsFormProvider.scala [hunk 2]
```diff
@@ -40,6 +40,6 @@ class AgentCompanyDetailsFormProvider extends Mappings {
         "agentCompanyDetails.error.agentPostalCode.required",
         "agentCompanyDetails.error.agentPostalCode.gb"
       )
-    )(AgentCompanyDetails.apply)(AgentCompanyDetails.unapply)
+    )(AgentCompanyDetails.apply)(o => Some(Tuple.fromProductTyped(o)))
```
[the remark this answers is not in the document]
Comment (SachaMcCormack) [reply app/forms/AgentCompanyDetailsFormProvider.scala:43]: In Scala 3, the signature of unapply has changed and causes an error.  On the Scala 3 documentation, it states to change it to o => Some(Tuple.fromProductTyped(o))

File: app/controllers/auth/AuthController.scala [hunk 3]
```diff
@@ -27,7 +27,7 @@ import javax.inject.Inject
 import scala.annotation.nowarn
 import scala.concurrent.ExecutionContext
 
-@nowarn("cat=unused")
+@nowarn("msg=unused")
```
[the remark this answers is not in the document]
Comment (SachaMcCormack) [reply app/controllers/auth/AuthController.scala:30]: image

Screenshot from 2024-11-07 09-00-18

Needed to change in the scalacOptions in build.sbt to msg so did the same here.

File: app/models/package.scala [hunk 4]
```diff
@@ -104,7 +104,7 @@ package object models {
           JsSuccess(JsArray(updatedJsArray))
         case valueToRemoveFrom: JsArray                                                         =>
           JsError(s"array index out of bounds: $index, $valueToRemoveFrom")
-        case _                                                                                  => JsError(s"cannot set an index on $valueToRemoveFrom")
+        case null                                                                               => JsError(s"cannot set an index on $valueToRemoveFrom")
```
[the remark this answers is not in the document]
Comment (SachaMcCormack) [reply app/models/package.scala:107]: This was producing a warning but will change back if won't catch certain instances

File: app/connectors/EmailConnector.scala [hunk 5]
```diff
@@ -29,13 +30,13 @@ import scala.concurrent.{ExecutionContext, Future}
 class EmailConnector @Inject() (
   httpClient: HttpClientV2,
   appConfig: FrontendAppConfig
-)(implicit ec: ExecutionContext) {
+)(using ec: ExecutionContext) {
 
   def sendEmail(
     emailRequest: EmailRequest
   )(implicit hc: HeaderCarrier): Future[HttpResponse] =
```
[the remark this answers is not in the document]
Comment (SachaMcCormack) [reply app/connectors/EmailConnector.scala]: For an unknown reason, changing implicit to given causes a syntax error in the EmailServiceSpec.scala file

File: test/controllers/actions/AuthActionSpec.scala [hunk 6]
```diff
@@ -68,8 +68,8 @@ class AuthActionSpec extends SpecBase {
     when(mockAuthConnector.authorise[RetrievalType](any(), any())(any(), any()))
```
Comment (lmugnai) [inline test/controllers/actions/AuthActionSpec.scala:68]: Can we change this to:

when(mockAuthConnector.authorise[RetrievalType](any(), any())(any(), any()))
  .thenReturn(
    Future.successful(
      internalId.composeRetrievals(enrolments(enrolmentKey)).composeRetrievals(Some(affinityGroup)).composeRetrievals(Some(credentialRole))
    )
  )

**Review by lmugnai — APPROVED** [review]

**Pull request merged**